contracts/contracts/token/OUSD.sol

```diff
@@ -516,47 +547,59 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
     }
 
     function _rebaseOptIn(address _account) internal {
-        require(_isNonRebasingAccount(_account), "Account has not opted out");
-
-        // Convert balance into the same amount at the current exchange rate
-        uint256 newCreditBalance = _creditBalances[_account]
-            .mul(_rebasingCreditsPerToken)
-            .div(_creditsPerToken(_account));
-
-        // Decreasing non rebasing supply
-        nonRebasingSupply = nonRebasingSupply.sub(balanceOf(_account));
+        require(_isNonRebasingAccount(_account), "Account must be non-rebasing");
+        RebaseOptions state = rebaseState[_account];
+        require(
+            state == RebaseOptions.StdNonRebasing ||
+                state == RebaseOptions.NotSet,
+            "Only standard non-rebasing accounts can opt out"
+        );
 
-        _creditBalances[_account] = newCreditBalance;
+        uint256 balance = balanceOf(msg.sender);
 
-        // Increase rebasing credits, totalSupply remains unchanged so no
-        // adjustment necessary
-        _rebasingCredits = _rebasingCredits.add(_creditBalances[_account]);
+        // Account
+        rebaseState[msg.sender] = RebaseOptions.StdRebasing;
+        alternativeCreditsPerToken[msg.sender] = 0;
+        _creditBalances[msg.sender] = _balanceToRebasingCredits(balance);
 
-        rebaseState[_account] = RebaseOptions.OptIn;
+        // Globals
+        nonRebasingSupply -= balance;
+        _rebasingCredits += _creditBalances[msg.sender];
 
-        // Delete any fixed credits per token
-        delete nonRebasingCreditsPerToken[_account];
         emit AccountRebasingEnabled(_account);
     }
 
-    /**
-     * @dev Explicitly mark that an address is non-rebasing.
-     */
     function rebaseOptOut() public nonReentrant {
-        require(!_isNonRebasingAccount(msg.sender), "Account has not opted in");
+        _rebaseOptOut(msg.sender);
+    }
 
-        // Increase non rebasing supply
-        nonRebasingSupply = nonRebasingSupply.add(balanceOf(msg.sender));
-        // Set fixed credits per token
-        nonRebasingCreditsPerToken[msg.sender] = _rebasingCreditsPerToken;
+    function _rebaseOptOut(address _account) internal {
+        RebaseOptions state = rebaseState[_account];
+        if (state == RebaseOptions.StdRebasing) {
+            require(
+                alternativeCreditsPerToken[_account] == 0,
+                "Account must be rebasing"
+            );
+        }
 
-        // Decrease rebasing credits, total supply remains unchanged so no
-        // adjustment necessary
-        _rebasingCredits = _rebasingCredits.sub(_creditBalances[msg.sender]);
+        require(
+            state == RebaseOptions.StdRebasing || state == RebaseOptions.NotSet,
+            "Only standard rebasing accounts can opt out"
+        );
```

Comment:
`state == RebaseOptions.StdRebasing` check is above this in a if statement. Could do something like
```
if (state == StdRebasing) {
...
} else {
require(state == NotSet, "");
}

Reply:
yeah was thinking to do such a change, but it seemed a little worse aesthetically. Happy to turn it into that if anyone has a strong opinion.

Reply:
See note in https://github.com/OriginProtocol/origin-dollar/pull/2298/files/45f476bad3dfdfa3c80ab2e7245c9edc2527140f..9a29c45205045cfdf4a27cb846405b1cb9e76834#r1832682477